### PR TITLE
JAMES-2861 ImportEml implementation

### PR DESCRIPTION
**Pull request by duongtq**
- ImportEml implementation.
- Documentation in progress.

### CODE REVIEW

File: server/protocols/webadmin/webadmin-data/pom.xml [hunk 1]
```diff
@@ -34,7 +34,28 @@
     <dependencies>
         <dependency>
             <groupId>${james.groupId}</groupId>
-            <artifactId>event-sourcing-event-store-memory</artifactId>
+            <artifactId>apache-james-mailbox-api</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>${james.groupId}</groupId>
+            <artifactId>apache-james-mailbox-memory</artifactId>
```
Comment (trantienduchn) [inline server/protocols/webadmin/webadmin-data/pom.xml]: be respect alphabetically dependencies ordering

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java [hunk 2]
```diff
@@ -744,4 +743,4 @@ public boolean hasChildren(MailboxPath mailboxPath, MailboxSession session) thro
         Mailbox mailbox = mapper.findMailboxByPath(mailboxPath);
         return mapper.hasChildren(mailbox, session.getPathDelimiter());
     }
-}
+}
```
Comment (trantienduchn) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java]: please avoid this kind of change

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxPath.java [hunk 3]
```diff
@@ -84,7 +84,7 @@ public String getNamespace() {
      * Get the name of the user who owns the mailbox. This can be null e.g. for
      * shared mailboxes.
      * 
-     * @return The username
+     * @return The usernameMailboxPath
```
Comment (trantienduchn) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxPath.java]: this change is not related, please revert the code.

File: server/protocols/webadmin/webadmin-data/pom.xml [hunk 4]
```diff
@@ -149,6 +170,24 @@
             <groupId>org.slf4j</groupId>
             <artifactId>slf4j-api</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.apache.james</groupId>
```
Comment (trantienduchn) [inline server/protocols/webadmin/webadmin-data/pom.xml]: use `${james.groupId}`

idem for some below dependencies

File: server/protocols/webadmin/webadmin-data/pom.xml [hunk 5]
```diff
@@ -149,6 +170,24 @@
...
+            <artifactId>james-server-mailbox-jmx</artifactId>
```
Comment (trantienduchn) [inline server/protocols/webadmin/webadmin-data/pom.xml]: do you really need this dependency?

File: server/protocols/webadmin/webadmin-data/pom.xml [hunk 6]
```diff
@@ -149,6 +170,24 @@
...
+            <artifactId>apache-james-mailbox-store</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.james</groupId>
+            <artifactId>apache-james-mailbox-memory</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.james</groupId>
+            <artifactId>event-sourcing-event-store-memory</artifactId>
```
Comment (trantienduchn) [inline server/protocols/webadmin/webadmin-data/pom.xml]: do you really need this dependency?

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/EmlService.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        mailbox.appendMessage(MessageManager.AppendCommand.builder()
+            .recent()
+            .build(new ByteArrayInputStream(request.bodyAsBytes())), session);
+
+    }
+
+    public MailboxSession getMailboxSession() throws MailboxException {
+        MailboxSession session;
+        try {
+            session = mailboxManager.createSystemSession("james@linagora.com");
```
Comment (trantienduchn) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/EmlService.java]: this user doesn't relate much to the context. I propose `EmlService`

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/EmlService.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        } catch (BadCredentialsException e) {
+            throw ErrorResponder.builder()
+                .statusCode(HttpStatus.BAD_REQUEST_400)
+                .type(ErrorResponder.ErrorType.INVALID_ARGUMENT)
+                .message("Bad credential.")
```
Comment (trantienduchn) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/EmlService.java]: what do you mean when answering to client a message `Bad credential.`?

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/EmlService.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+                .haltError();
+        }
+        return session;
+    }
+
+    private MessageManager getMailboxFromMailboxIdAndSession(Request request) throws MailboxException {
```
Comment (trantienduchn) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/EmlService.java]: should has a simple name `retrieveMailbox()`

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/EmlService.java [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+        String mailboxIdString = request.params(MAILBOX_ID);
+        MailboxId mailboxId = mailboxIdFactory.fromString(mailboxIdString);
```
Comment (trantienduchn) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/EmlService.java]: 2 lines can be in one

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/EmlService.java [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+        MailboxSession session = getMailboxSession();
+
+        MessageManager mailbox;
+
+        try {
+            mailbox = mailboxManager.getMailbox(mailboxId, session);
```
Comment (trantienduchn) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/EmlService.java]: how about returning right here?

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/EmlService.java [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+        } catch (MailboxException e) {
+            throw ErrorResponder.builder()
+                .statusCode(HttpStatus.BAD_REQUEST_400)
+                .type(ErrorResponder.ErrorType.INVALID_ARGUMENT)
+                .message("Unable to create mailbox.")
```
Comment (trantienduchn) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/EmlService.java]: `Unable to create mailbox.` => `Unable to retrieve mailbox.`

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/EmlService.java [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+        MailboxId mailboxId = mailboxIdFactory.fromString(mailboxIdString);
+
+        MailboxSession session = getMailboxSession();
+
+        MessageManager mailbox;
+
+        try {
+            mailbox = mailboxManager.getMailbox(mailboxId, session);
+        } catch (MailboxException e) {
+            throw ErrorResponder.builder()
```
Comment (trantienduchn) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/EmlService.java]: I think sending a response should be the responsibility of the route not the service

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/EmlRoutesTest.java [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+
+        with()
+            .body(ClassLoader.getSystemResource("eml/mail.eml"))
+            .post("/" + dummyInbox.serialize() + "/messages")
+        .then()
+            .statusCode(HttpStatus.NO_CONTENT_204);
+    }
+
+    @Test
+    public void importEmlToMailBoxShouldAllowOtherFileExtensions() throws MailboxException {
```
Comment (trantienduchn) [inline server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/EmlRoutesTest.java]: It should reject other extensions IMO

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/EmlRoutesTest.java [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+
+        with()
+            .body(ClassLoader.getSystemResource("eml/mail.json"))
+            .post("/" + dummyInbox.serialize() + "/messages")
+        .then()
+            .statusCode(HttpStatus.NO_CONTENT_204);
+    }
+
+    @Test
+    public void importEmlToMailboxShouldAllowEmptyEmlPath() throws MailboxException {
```
Comment (trantienduchn) [inline server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/EmlRoutesTest.java]: what do you mean empty path? the route doesn't care the file path, it just cares about the content being uploaded

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/EmlRoutesTest.java [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+    public void importEmlToMailboxShouldReturnBadRequestWhenUsernameDoesNotMatch() throws MailboxException {
+
+        session = mailboxManager.createSystemSession("EML");
+        dummyInbox = mailboxManager
+            .createMailbox(MailboxPath.forUser("EML", "inbox"), session)
+            .get();
+
+        with()
+            .body(ClassLoader.getSystemResource("eml/mail.eml"))
+            .post("/" + dummyInbox.serialize() + "/messages")
```
Comment (trantienduchn) [inline server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/EmlRoutesTest.java]: this request is suppose to be success with every mailboxId

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/EmlRoutesTest.java [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+        .then()
+            .statusCode(HttpStatus.BAD_REQUEST_400)
+            .body("message", is("Unable to create mailbox."));
+    }
+
+    @Test
+    public void importEmlToMailboxShouldReturnNotFoundWhenPathIsNotCorrect() throws MailboxException {
```
Comment (trantienduchn) [inline server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/EmlRoutesTest.java]: again, the route doesn't care the file path...

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/dto/MailboxResponse.java [hunk 18]
```diff
@@ -19,15 +19,17 @@
...
 package org.apache.james.webadmin.dto;
 
+import org.apache.james.mailbox.model.MailboxId;
+
 public class MailboxResponse {
 
     private final String mailboxName;
 
-    public MailboxResponse(String mailboxName) {
+    public MailboxResponse(String mailboxName, MailboxId mailboxId) {
```
Comment (trantienduchn) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/dto/MailboxResponse.java]: how does this change relate to the content of the task?

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxPath.java (hunk 3 above)
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxPath.java]: Are you sure?

File: server/container/guice/protocols/webadmin-data/src/main/java/org/apache/james/modules/server/DataRoutesModules.java [hunk 19]
```diff
@@ -46,11 +47,12 @@ protected void configure() {
         routesMultibinder.addBinding().to(DomainMappingsRoutes.class);
         routesMultibinder.addBinding().to(ForwardRoutes.class);
         routesMultibinder.addBinding().to(GroupsRoutes.class);
+        routesMultibinder.addBinding().to(EmlRoutes.class);
```
Comment (chibenwa) [inline server/container/guice/protocols/webadmin-data/src/main/java/org/apache/james/modules/server/DataRoutesModules.java]: ImportEmlRoute

File: server/protocols/webadmin/webadmin-data/pom.xml [hunk 20]
```diff
@@ -34,7 +34,28 @@
     <dependencies>
         <dependency>
             <groupId>${james.groupId}</groupId>
-            <artifactId>event-sourcing-event-store-memory</artifactId>
+            <artifactId>apache-james-mailbox-api</artifactId>
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-data/pom.xml]: Why do I need mailbox related stuff in data projects? 

Looks like the EML route should be in webadmin-mailbox...

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/EmlRoutes.java [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+
+    @POST
+    @Path(IMPORT_EML_PATH)
+    @ApiOperation(value = "import an EML file and save it to a mailbox")
+    @ApiResponses(value = {
+        @ApiResponse(code = HttpStatus.NO_CONTENT_204, message = "EML file successfully imported"),
+        @ApiResponse(code = HttpStatus.BAD_REQUEST_400, message = "Invalid parameter")
+    })
+    public HaltException importEmlFileToMailbox(Request request, Response response) throws MailboxException {
+
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/EmlRoutes.java]: extra line

Comment (chibenwa) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/EmlRoutes.java]: extra line

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/EmlService.java [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+    private final MailboxManager mailboxManager;
+
+    @Inject
+    @VisibleForTesting
+    public EmlService(MailboxManager mailboxManager, MailboxId.Factory mailboxIdFactory) {
+        this.mailboxManager = mailboxManager;
+        this.mailboxIdFactory = mailboxIdFactory;
+    }
+
+    public void importEmlFileToMailbox(Request request) throws MailboxException {
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/EmlService.java]: Parsing of the request into domain objects is a responsability of the route, not of the service. (Hint: do not pass the Request here as a parameter)

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/EmlService.java [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+
+        MailboxSession session = getMailboxSession();
+        MessageManager mailbox = getMailboxFromMailboxIdAndSession(request);
+        mailbox.appendMessage(MessageManager.AppendCommand.builder()
+            .recent()
+            .build(new ByteArrayInputStream(request.bodyAsBytes())), session);
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/EmlService.java]: This looks.... Blocking. I'd prefer a call: `request.raw().getInputStream()` to be doing some streaming

**Comment by Arsnael** [comment]
```
[ERROR]     'dependencies.dependency.version' for org.apache.james:apache-james-mailbox-memory:jar is missing. @ org.apache.james:james-server-webadmin-data:[unknown-version], /james-project/server/protocols/webadmin/webadmin-data/pom.xml, line 39, column 21
```
?

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/EmlService.java (hunk 7 above)
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/EmlService.java]: You could be writing the return here to avoid reallocation

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/EmlService.java [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+
+    }
+
+    public MailboxSession getMailboxSession() throws MailboxException {
+        MailboxSession session;
+        try {
+            session = mailboxManager.createSystemSession("james@linagora.com");
+        } catch (BadCredentialsException e) {
+            throw ErrorResponder.builder()
+                .statusCode(HttpStatus.BAD_REQUEST_400)
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/EmlService.java]: That would be a 500 and it's the route responsibility to return HTTP exception, not of the service.

Here `throw new RuntimeException(xxx, e);` would be acceptable to me.

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/EmlService.java (hunk 7 above)
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/EmlService.java]: Let's use the service name as a  session name: `WebadminImportEML`

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/EmlService.java (hunk 11 above)
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/EmlService.java]: Avoid variable realocation by all means. Rely on method extraction when needed.

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/EmlRoutesTest.java [hunk 25]
```diff
@@ -0,0 +1,128 @@
+package org.apache.james.webadmin.routes;
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/EmlRoutesTest.java]: missing apachev2 license

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/EmlRoutesTest.java [hunk 26]
```diff
@@ -0,0 +1,128 @@
...
+import org.apache.james.webadmin.service.EmlService;
+import org.eclipse.jetty.http.HttpStatus;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import io.restassured.RestAssured;
+import io.restassured.filter.log.LogDetail;
+
+public class EmlRoutesTest {
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/EmlRoutesTest.java]: I don't see test inspecting the mailboxManager to see that the message was well created with the right content.

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/dto/MailboxResponse.java (hunk 18 above)
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/dto/MailboxResponse.java]: Where is that constructor field used?

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/EmlRoutesTest.java (hunk 17 above)
Comment (duongtq) [reply server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/EmlRoutesTest.java]: Not the file path, I mean the URL.

File: mailbox/memory/src/test/resources/file.json [hunk 27]
```diff
@@ -0,0 +1,22 @@
+{
```
Comment (chibenwa) [inline mailbox/memory/src/test/resources/file.json]: What's this?

File: mailbox/memory/src/test/resources/importEmlTest.eml [hunk 28]
```diff
@@ -0,0 +1,65 @@
...
+X-Spam-Status: No, score=-1.51 required=1 tests=[BAYES_00=-1.5,
+	T_RP_MATCHES_RCVD=-0.01] autolearn=ham
+Received: from mx2.minet.net ([IPv6:::ffff:192.168.102.26])
+	by localhost (spam.minet.net [::ffff:192.168.102.97]) (amavisd-new, port 10024)
+	with ESMTP id IeILbadS9lo5 for <benwa@minet.net>;
+	Thu,  4 Jun 2015 09:23:42 +0000 (UTC)
+Received-SPF: Pass (sender SPF authorized) identity=mailfrom; client-ip=140.211.11.3; helo=mail.apache.org; envelope-from=server-dev-return-56862-benwa=minet.net@james.apache.org; receiver=benwa@minet.net
+Received: from mail.apache.org (hermes.apache.org [140.211.11.3])
+	by mx2.minet.net (Postfix) with SMTP id CDE83A1C7FC
+	for <benwa@minet.net>; Thu,  4 Jun 2015 11:23:40 +0200 (CEST)
```
Comment (chibenwa) [inline mailbox/memory/src/test/resources/importEmlTest.eml]: Please impersonate this (no "real " mail address)

File: server/protocols/webadmin/webadmin-data/pom.xml [hunk 29]
```diff
@@ -7,16 +7,15 @@
     to you under the Apache License, Version 2.0 (the
     "License"); you may not use this file except in compliance
     with the License. You may obtain a copy of the License at
-
     http://www.apache.org/licenses/LICENSE-2.0
-
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-data/pom.xml]: Why?

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/service/ImportEmlService.java [hunk 30]
```diff
@@ -0,0 +1,56 @@
...
+        MailboxSession webAdminImportEMLSession = mailboxManager.createSystemSession("EmlService");
+        return webAdminImportEMLSession;
+    }
+
+    public MessageManager retrieveMailbox(String mailboxIdParameter) throws MailboxException {
+        MailboxId mailboxId = mailboxIdFactory.fromString(mailboxIdParameter);
+        MailboxSession session = getMailboxSession();
+
+        MessageManager mailbox = mailboxManager.getMailbox(mailboxId, session);
+        return mailbox;
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/service/ImportEmlService.java]: `return mailboxManager.getMailbox(mailboxId, session);` directly...

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/service/ImportEmlService.java [hunk 31]
```diff
@@ -0,0 +1,56 @@
...
+    @Inject
+    @VisibleForTesting
+    public ImportEmlService(MailboxManager mailboxManager, MailboxId.Factory mailboxIdFactory) {
+        this.mailboxManager = mailboxManager;
+        this.mailboxIdFactory = mailboxIdFactory;
+    }
+
+    public MailboxSession getMailboxSession() throws MailboxException {
+        MailboxSession webAdminImportEMLSession = mailboxManager.createSystemSession("EmlService");
+        return webAdminImportEMLSession;
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/service/ImportEmlService.java]: return directly. Is method extraction justified here?

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ImportEmlRoutesTest.java [hunk 32]
```diff
@@ -0,0 +1,178 @@
...
+        RestAssured.requestSpecification = WebAdminUtils.buildRequestSpecification(webAdminServer)
+            .setBasePath(ImportEmlRoutes.BASE_PATH)
+            .log(LogDetail.ALL)
+            .build();
+    }
+
+    @AfterEach
+    public void stop() {
+        mailboxManager.logout(session, false);
+        mailboxManager.endProcessingRequest(session);
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ImportEmlRoutesTest.java]:
```
        mailboxManager.logout(session, false);
        mailboxManager.endProcessingRequest(session);
```

not needed here. (memory mailbox manager)

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ImportEmlRoutesTest.java [hunk 33]
```diff
@@ -0,0 +1,178 @@
...
+            .post("/" + dummyInbox.serialize() + "/messages")
+        .then()
+            .statusCode(HttpStatus.NO_CONTENT_204);
+
+        assertThat(mailboxManager.getMailbox(MailboxPath.forUser("EmlService", "inbox"), session)
+            .getMessages(MessageRange.all(), FetchGroupImpl.FULL_CONTENT, session))
+            .toIterable()
+            .hasSize(1)
+            .allSatisfy(Throwing.consumer(result -> assertThat(result
+                                                        .getBody()
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ImportEmlRoutesTest.java]: ident?

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ImportEmlRoutesTest.java [hunk 34]
```diff
@@ -0,0 +1,178 @@
...
+                                                        .getInputStream())
+                                                        .hasSameContentAs(ClassLoader.getSystemResource("importEmlTest.eml").openStream())));
+    }
+
+    @Test
+    public void importEmlToMailBoxShouldAllowOtherFileExtensions() throws Exception {
+
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ImportEmlRoutesTest.java]: Get rid of these extra lines

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ImportEmlRoutesTest.java [hunk 35]
```diff
@@ -0,0 +1,178 @@
...
+        session = mailboxManager.createSystemSession("EmlService");
+        dummyInbox = mailboxManager
+            .createMailbox(MailboxPath.forUser("EmlService", "inbox"), session)
+            .get();
+
+        given()
+            .body(IOUtils.toString(ClassLoader.getSystemResource("file.json").openStream()))
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ImportEmlRoutesTest.java]: IOUtils.toString should always take a charset (UTF8)

Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ImportEmlRoutesTest.java]: Use `ClassLoader.getSystemResourceAsString(...)`

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ImportEmlRoutesTest.java [hunk 36]
```diff
@@ -0,0 +1,178 @@
...
+            .statusCode(HttpStatus.NO_CONTENT_204);
+
+        assertThat(mailboxManager.getMailbox(MailboxPath.forUser("EmlService", "inbox"), session)
+            .getMessages(MessageRange.all(), FetchGroupImpl.FULL_CONTENT, session))
+            .toIterable()
+            .hasSize(1)
+            .allSatisfy(Throwing.consumer(result -> assertThat(result
+                                                        .getBody()
+                                                        .getInputStream())
+                                                        .hasSameContentAs(ClassLoader.getSystemResource("").openStream())));
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ImportEmlRoutesTest.java]: What is `""`?

**Comment by chibenwa** [comment]
Globally good contribution, some minor issues to fix IMO

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ImportEmlRoutesTest.java [hunk 37]
```diff
@@ -0,0 +1,178 @@
...
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import com.github.fge.lambdas.Throwing;
+
+import io.restassured.RestAssured;
+import io.restassured.filter.log.LogDetail;
+
+public class ImportEmlRoutesTest {
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ImportEmlRoutesTest.java]: please add : 
* a test with a big file that is supposed to work
* a test with a too big file
* a test with a invalid file format

Comment (duongtq) [reply server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ImportEmlRoutesTest.java]: IMO only two last requirements should be added. If the routes can handle an extremely big file, they definitely can handle big file, too.

Comment (duongtq) [reply server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ImportEmlRoutesTest.java]: IMO only the last requirement should be added. If the routes can handle an extremely big file, they definitely can handle big file, too.
File format should be eml.

Comment (mbaechler) [reply server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ImportEmlRoutesTest.java]: If you don't have a test, you can't know whether the application will crash, if there's a way to configure the limit, what the limit is, at which moment you are rejecting a file above the limit, etc.
So, please add those tests.

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/ImportEmlRoutes.java [hunk 38]
```diff
@@ -83,33 +83,33 @@ public void define(Service service) {
         @ApiResponse(code = HttpStatus.NO_CONTENT_204, message = "EML file successfully imported"),
         @ApiResponse(code = HttpStatus.BAD_REQUEST_400, message = "Invalid parameter")
     })
-    public HaltException importEmlFileToMailbox(Request request, Response response) throws MailboxException {
-
+    public HaltException importEmlFileToMailbox(Request request, Response response) throws Exception {
+        final byte[] fileSize = request.body().getBytes();
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/ImportEmlRoutes.java]: so if I upload a 1GiB file, you will start by putting it in memory to tell me it's too big?
I'll probably break the server long before I receive the message.

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ImportEmlRoutesTest.java (hunk 37 above)
Comment (duongtq) [reply server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ImportEmlRoutesTest.java]: The ticket only requires the route to import the body of an EML file, it does not require format validation.

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/ImportEmlRoutes.java [hunk 39]
```diff
@@ -83,33 +83,33 @@ public void define(Service service) {
...
                 .build(request.body()), session);
-        } catch (MailboxException e) {
+
+        if (fileSize.length > MAXIMUM_SIZE_IN_BYTE) {
             throw ErrorResponder.builder()
-                .statusCode(HttpStatus.BAD_REQUEST_400)
-                .type(ErrorResponder.ErrorType.INVALID_ARGUMENT)
-                .message("Invalid file extension.")
-                .haltError();
+            .statusCode(HttpStatus.INTERNAL_SERVER_ERROR_500)
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/ImportEmlRoutes.java]: you basically never have to use INTERNAL_SERVER_ERROR_500, it's for unhandled problems.
Please learn HTTP codes to choose them wisely (HTTP 413 here)

**Comment by duongtq** [comment]
Progress (6): 2.7/4.4 MB | 1.8/2.9 MB | 1.9/13[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.7.0:compile (default-compile) on project metrics-api: Fatal error compiling: invalid flag: --release -> [Help 1]

test this please

**Comment by chibenwa** [comment]
@duongtq you need to rebase on master to use latest compilation image

**Comment by duongtq** [comment]
test this please

**Comment by duongtq** [comment]
[46e6ed514073bfc8cd23382022a853564252bd3d] ERROR StatusLogger Log4j2 could not find a logging implementation. Please add log4j-core to the classpath. Using SimpleLogger to log to the console...

test this please

**Comment by duongtq** [comment]
test this please

**Comment by chibenwa** [comment]
Many merge conflict, inactive for very long, opened on linagora repo and not the ASF one...

Reopen on ASF repository if relevant.
